File: python/jupyter notebook/3DMotionPlanning/motion_planning.py
import argparse
import time
import msgpack
from enum import Enum, auto

import numpy as np

from planning_utils import a_star, heuristic, create_grid, a_star_for_graph
from udacidrone import Drone
from udacidrone.connection import MavlinkConnection
from udacidrone.messaging import MsgID
from udacidrone.frame_utils import global_to_local
from sampling import Sampler
from shapely.geometry import LineString
import networkx as nx
from sklearn.neighbors import KDTree
import matplotlib.pyplot as plt


class States(Enum):
    MANUAL = auto()
    ARMING = auto()
    TAKEOFF = auto()
    WAYPOINT = auto()
    LANDING = auto()
    DISARMING = auto()
    PLANNING = auto()


class MotionPlanning(Drone):

    def __init__(self, connection):
        super().__init__(connection)

        self.target_position = np.array([0.0, 0.0, 0.0])
        self.waypoints = []
        self.in_mission = True
        self.check_state = {}
        self.polygons = []

        # initial state
        self.flight_state = States.MANUAL

        # register all your callbacks here
        self.register_callback(MsgID.LOCAL_POSITION, self.local_position_callback)
        self.register_callback(MsgID.LOCAL_VELOCITY, self.velocity_callback)
        self.register_callback(MsgID.STATE, self.state_callback)

    def local_position_callback(self):
        if self.flight_state == States.TAKEOFF:
            if -1.0 * self.local_position[2] > 0.95 * self.target_position[2]:
                self.waypoint_transition()
        elif self.flight_state == States.WAYPOINT:
            if np.linalg.norm(self.target_position[0:2] - self.local_position[0:2]) < 1.0:
                if len(self.waypoints) > 0:
                    self.waypoint_transition()
                else:
                    if np.linalg.norm(self.local_velocity[0:2]) < 1.0:
                        self.landing_transition()

    def velocity_callback(self):
        if self.flight_state == States.LANDING:
            if self.global_position[2] - self.global_home[2] < 0.1:
                if abs(self.local_position[2]) < 0.01:
                    self.disarming_transition()

    def state_callback(self):
        if self.in_mission:
            if self.flight_state == States.MANUAL:
                self.arming_transition()
            elif self.flight_state == States.ARMING:
                if self.armed:
                    self.plan_path()
            elif self.flight_state == States.PLANNING:
                self.takeoff_transition()
            elif self.flight_state == States.DISARMING:
                if ~self.armed & ~self.guided:
                    self.manual_transition()

    def arming_transition(self):
        self.flight_state = States.ARMING
        print("arming transition")
        self.arm()
        self.take_control()

    def takeoff_transition(self):
        self.flight_state = States.TAKEOFF
        print("takeoff transition")
        self.takeoff(self.target_position[2])

    def waypoint_transition(self):
        self.flight_state = States.WAYPOINT
        print("waypoint transition")
        self.target_position = self.waypoints.pop(0)
        print('target position', self.target_position)
        self.cmd_position(self.target_position[0], self.target_position[1], self.target_position[2], self.target_position[3])

    def landing_transition(self):
        self.flight_state = States.LANDING
        print("landing transition")
        self.land()

    def disarming_transition(self):
        self.flight_state = States.DISARMING
        print("disarm transition")
        self.disarm()
        self.release_control()

    def manual_transition(self):
        self.flight_state = States.MANUAL
        print("manual transition")
        self.stop()
        self.in_mission = False

    def send_waypoints(self):
        print("Sending waypoints to simulator ...")
        data = msgpack.dumps(self.waypoints)
        self.connection._master.write(data)

    def plan_path(self):
        self.flight_state = States.PLANNING

        print("Searching for a path ...")
        TARGET_ALTITUDE = 5
        SAFETY_DISTANCE = 5

        self.target_position[2] = TARGET_ALTITUDE

        # TODO: read lat0, lon0 from colliders into floating point values
        data = np.genfromtxt('colliders.csv', delimiter=',', dtype='str', max_rows=1)
        _, lat0 = data[0].strip().split(' ')
        _, lon0 = data[1].strip().split(' ')
        # TODO: set home position to (lon0, lat0, 0)
        self.set_home_position(float(lon0), float(lat0), 0)
        # TODO: retrieve current global position
        curr_glb_pos = [self._longitude, self._latitude, self._altitude]
        # TODO: convert to current local position using global_to_local()
        self._north, self._east, self._down = global_to_local(curr_glb_pos, self.global_home)
        print('global home {0}, position {1}, local position {2}'.format(self.global_home, self.global_position,
                                                                         self.local_position))
        # Read in obstacle map
        data = np.loadtxt('colliders.csv', delimiter=',', dtype='Float64', skiprows=2)

        sampler = Sampler(data, SAFETY_DISTANCE)
        self.polygons = sampler._polygons
        nodes = sampler.sample(300)
        print('nodes_len: ', len(nodes))

        g = self.create_graph(nodes, 10)
        print('graph_edgs: ', len(g.edges))
        start = self.local_position
        goal = global_to_local([-122.396428, 37.795128, TARGET_ALTITUDE], self.global_home)

        start = self.find_closest_node(g.nodes, start)
        goal = self.find_closest_node(g.nodes, goal)
        path, cost = a_star_for_graph(g, heuristic, start, goal)
        print('a_star_path: ', path)
        path = self.prune_path(path)
        print('prune_path: ' ,path)
        if len(path) > 0:
            # Convert path to waypoints
            waypoints = [[p[0], p[1], TARGET_ALTITUDE, 0] for p in path]
            # Set self.waypoints
            self.waypoints = waypoints
            # TODO: send waypoints to sim (this is just for visualization of waypoints)
            self.send_waypoints()

    def start(self):
        #self.start_log("Logs", "NavLog.txt")

        print("starting connection")
        self.connection.start()

        # Only required if they do threaded
        # while self.in_mission:
        #    pass

        #Sself.stop_log()

    def can_connect(self, n1, n2):
        # Determine whether there is a feasible route between two nodes
        l = LineString([n1, n2])
        h = min(n1[2], n2[2])
        for p in self.polygons:
            if p.crosses(l) and p.height >= h:
                return False
        return True

    def create_graph(self, nodes, k):
        # Create a feasible roadmap, use KDTree to find the k nearest neighbors, detect collisions, and build a roadmap
        g = nx.Graph()
        tree = KDTree(nodes)
        for n1 in nodes:
            # for each node connect try to connect to k nearest nodes
            idxs = tree.query([n1], k, return_distance=False)[0]
            for idx in idxs:
                n2 = nodes[idx]
                if n2 == n1:
                    continue
                if self.can_connect(n1, n2):
                    g.add_edge(n1, n2, weight=1)
        return g

    def bresenham_check(self, p1, p2):
        # Use bresenham algorithm to determine whether there is a feasible route between p1 and p2
        x1, y1, _ = p1
        x2, y2, _ = p2
        cells = []

        dx = x2 - x1
        dy = y2 - y1
        x = x1
        y = y1
        eps = 0
        if dx == 0:
            return self.can_connect(p1, p2)

        m = dy / dx
        # region 1
        if 0 <= m <= 1 and x1 < x2:
            for x in range(x1, x2):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dy
                if ((eps << 1) >= dx):
                    y += 1
                    eps -= dx
        # region 2
        elif m > 1 and y1 < y2:
            for y in range(y1, y2):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dx
                if ((eps << 1) >= dy):
                    x += 1
                    eps -= dy
        # region 3
        elif m < -1 and y1 < y2:
            for y in range(y1, y2):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dx
                if ((eps << 1) <= -dy):
                    x -= 1
                    eps += dy
        # region 4
        elif -1 <= m <= 0 and x2 < x1:
            for x in range(x1, x2, -1):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dy
                if ((eps << 1) >= -dx):
                    y += 1
                    eps += dx
        # region 5
        elif 0 < m <= 1 and x2 < x1:
            for x in range(x1, x2, -1):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dy
                if ((eps << 1) <= dx):
                    y -= 1
                    eps -= dx
        # region 6
        elif m > 1 and y2 < y1:
            for y in range(y1, y2, -1):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dx
                if ((eps << 1) <= dy):
                    x -= 1
                    eps -= dy
        # region 7
        elif m < -1 and y2 < y1:
            for y in range(y1, y2, -1):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dx
                if ((eps << 1) >= -dy):
                    x += 1
                    eps += dy
        # region 8
        elif -1 <= m < 0 and x1 < x2:
            for x in range(x1, x2):
                if not self.can_connect(p1, (x, y, p2[2])):
                    return False
                cells.append((x, y))
                eps += dy
                if ((eps << 1) <= -dx):
                    y -= 1
                    eps += dx

        return True

    def prune_path(self, path):
        # Remove redundant nodes in the path
        pruned_path = [p for p in path]
        i = 0
        while i < len(pruned_path) - 2:
            p1 = pruned_path[i]
            p2 = pruned_path[i + 1]
            p3 = pruned_path[i + 2]
            # First check if p2 can be removed, if it can be checked whether p3 can be removed, if both are satisfied, remove p2
            if self.bresenham_check(p1, p2) and self.bresenham_check(p1, p3):
                pruned_path.remove(pruned_path[i+1])
            else:
                i += 1
        return pruned_path

    def find_closest_node(self, points, point):
        # Find the node closest to a given point in the graph
        start_min_dist = np.linalg.norm(np.array(point) - np.array(points), axis=1).argmin()
        near_node = np.array(points)[start_min_dist]
        return tuple(near_node)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--port', type=int, default=5760, help='Port number')
    parser.add_argument('--host', type=str, default='127.0.0.1', help="host address, i.e. '127.0.0.1'")
    args = parser.parse_args()

    conn = MavlinkConnection('tcp:{0}:{1}'.format(args.host, args.port), timeout=60)
    drone = MotionPlanning(conn)
    time.sleep(1)

    drone.start()

File: drone/MotionPlanning.cpp
#include "MotionPlanning.hpp"
#include "SearchAlgorithm.hpp"

MotionPlanning::MotionPlanning(MavlinkConnection* conn) : Drone(conn)
{
    // register all your callbacks here
    register_callback(MessageIDs::LOCAL_POSITION, ((void (Drone::*)()) & MotionPlanning::local_position_callback));
    register_callback(MessageIDs::LOCAL_VELOCITY, ((void (Drone::*)()) & MotionPlanning::velocity_callback));
    register_callback(MessageIDs::STATE, ((void (Drone::*)()) & MotionPlanning::state_callback));
    register_callback(MessageIDs::COMMANDACKRESULT, ((void (Drone::*)()) & MotionPlanning::command_ack_callback));
}

void MotionPlanning::local_position_callback()
{
    if (flight_state == States::TAKEOFF)
    {
        if (abs(abs(local_position()[2]) - abs(target_position[2])) < 0.1)
        {
            waypoint_transition();
        }
    }
    else if (flight_state == States::WAYPOINT)
    {
        if ((target_position - local_position()).mag() < 1.0)
        {
            if (all_waypoints.size() > 0)
            {
                waypoint_transition();
            }
            else if (local_velocity().mag() < 1.0)
            {
                landing_transition();
            }
        }
    }
}

void MotionPlanning::velocity_callback()
{
    if (flight_state == States::LANDING)
    {
        if ((global_position()[2] - global_home()[2] < 0.1)
            && (abs(local_position()[2]) < 0.01))
        {
            disarming_transition();
        }
    }
}

void MotionPlanning::state_callback()
{
    if (in_mission)
    {
        if (flight_state == States::MANUAL)
        {
            if (!in_planning)
            {
                new thread(&MotionPlanning::plan_path, this);
                in_planning = true;
            }
        }
        else if (flight_state == States::ARMING && armed())
        {
            takeoff_transition();
        }
        else if (flight_state == States::PLANNING)
        {
            arming_transition();
        }
        else if (flight_state == States::DISARMING && !armed() && !guided())
        {
            manual_transition();
        }
    }
}

void MotionPlanning::arming_transition()
{
    flight_state = States::ARMING;
    cout << "arming transition\r\n" << endl;
    arm();
}

void MotionPlanning::takeoff_transition()
{
    cout << "takeoff transition\r\n" << endl;
    local_position().print();
    float target_altitude = local_position()[2] + 10.0;
    target_position[2] = target_altitude;
    takeoff(/*target_position[2]*/);
    flight_state = States::TAKEOFF;
}

void MotionPlanning::waypoint_transition()
{
    flight_state = States::WAYPOINT;
    cout << "waypoint transition" << endl;
    if (all_waypoints.size() <= 0)
    {
        return;
    }
    target_position = all_waypoints.front();
    all_waypoints.pop();
    target_position.print("target_position");
    cmd_position(target_position[0], target_position[1], target_position[2], 0.0);
}

void MotionPlanning::landing_transition()
{
    flight_state = States::LANDING;
    cout << "landing transition" << endl;
    land();
}

void MotionPlanning::disarming_transition()
{
    flight_state = States::DISARMING;
    cout << "disarm transition" << endl;
    disarm();
}

void MotionPlanning::manual_transition()
{
    flight_state = States::MANUAL;
    cout << "manual transition" << endl;
    stop();
    in_mission = false;
}

void MotionPlanning::start_drone()
{
    cout << "starting connection" << endl;
    start();
}

void MotionPlanning::find_closest_node(vector<V3F> nodes, V3F p, V3F& p_min)
{
    int i, index = 0;
    float dist, min_dist = 0;
    for (i = 0; i < nodes.size(); i++) {
        dist = p.distance(nodes[i]);
        if (i == 0 || dist < min_dist) {
            min_dist = dist;
            index = i;
        }
    }
    if (i > 0) {
        p_min = nodes[index];
    }
}

void MotionPlanning::send_waypoints(vector<V3F> points)
{
    // serialize the object into the buffer.
    // any classes that implements write(const char*,size_t) can be a buffer.
    cout << "Sending waypoints to simulator ..." << endl;
    vector<MSGPoint> msgPoint;
    for (size_t i = 0; i < points.size(); i++)
    {
        msgPoint.push_back(points[i]);
    }
    msgpack::sbuffer buffer;
    msgpack::pack(buffer, msgPoint);
    getConnection()->getMaster()->write(buffer.data(), buffer.size());
}

void MotionPlanning::plan_path()
{
    cout << "Searching for a path ..." << endl;
    float target_altitude = 5;
    float safety_distance = 5;

    target_position[2] = target_altitude;
    string path = "../../data/colliders.csv";
#ifdef WIN32
    path = "../../../data/colliders.csv";
#endif
    FreeData<float> data(path, ",");
    float lat0 = data.getLat();
    float lon0 = data.getLon();
    //set_home_position(lon0, lat0, 0);
    _update_local_position(global_to_local(global_position(), global_home()));

    data.extract_polygons(safety_distance);
    data.sample(1000);
    data.create_graph(10);
    V3F start = local_position();
    V3F goal({ 10, 10, 10 });
    //V3F goal = global_to_local({ -122.396428, 37.795128, target_altitude }, global_home());
    cout << "start and goal" << endl;
    start.print();
    goal.print();
    vector<FreeEdge<float, 3>> edges;
    vector<V3F> nodes;
    FreeGraph<float, 3> graph = data.getGraph();
    graph.getAllNodesAndEdges(nodes, edges);

    V3F start_new, goal_new;
    find_closest_node(nodes, start, start_new);
    find_closest_node(nodes, goal, goal_new);
    cout << "startnew and goalnew" << endl;
    start_new.print();
    goal_new.print();

    GirdCellType start_grid(start_new);
    GirdCellType goal_grid(goal_new);
    SearchAlgorithm a_start_graph(start_grid, goal_grid);
    a_start_graph.a_start_graph(graph);
    cout << "star graph end" << endl;
    vector<V3F> path_points_prune;
    vector<V3F> path_points = a_start_graph.get_path_points();
    cout << "get path points end" << path_points.size() << endl;
    if (path_points.size() > 0)
    {
        a_start_graph.prune_path_by_collinearity(path_points, path_points_prune);
        cout << "prune end" << endl;
        send_waypoints(path_points_prune);
        for (int i = 0; i < path_points_prune.size(); i++) {
            path_points_prune[i].print();
            all_waypoints.push({ path_points_prune[i][0], path_points_prune[i][1], path_points_prune[i][2] });
        }
        flight_state = States::PLANNING;
    }
    else
    {
        in_planning = false;
    }
    //V3F cp = local_position();
    //all_waypoints.push(cp + V3F({ 10.0, 0.0, 10 }));
    //all_waypoints.push(cp + V3F({ 10.0, 10.0, 10 }));
    //all_waypoints.push(cp + V3F({ 0.0, 10.0, 10 }));
    //all_waypoints.push(cp + V3F({ 0.0, 0.0, 10 }));
    //flight_state = States::PLANNING;
}

void MotionPlanning::command_ack_callback()
{
    //if (!m_bControlStatus && m_bTakeoffed)
    //{
    //    cout << "cmd offboard on" << endl;
    //    cmd_position(local_position()[0], local_position()[1], -abs(target_position[2]), 0);
    //    getConnection()->make_command_flight_mode(FlightMode::Offboard);
    //    m_bControlStatus = true;
    //}
}